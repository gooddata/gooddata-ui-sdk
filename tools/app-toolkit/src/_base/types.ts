// (C) 2021-2022 GoodData Corporation

import { OptionValues } from "commander";
import isEmpty from "lodash/isEmpty";

/**
 * Program & command level options are grabbed from the commander and sent together to the action.
 */
export type ActionOptions = {
    /**
     * Program-level options. These are the top level options defined directly using `program.option(...)`
     */
    programOpts: OptionValues;

    /**
     * Command-level options. These are options defined via `program.command(..).option(..)`
     */
    commandOpts: OptionValues;
};

/**
 * Supported languages to use in application projects bootstrapped by the CLI.
 */
export type TargetAppLanguage = "ts" | "js";

/**
 * Supported package managers that the tool can use to auto-install dependencies for the boostrapped projects.
 */
export type SupportedPackageManager = "npm" | "yarn";

/**
 * Supported templates that can be bootstrapped by the CLI.
 */
export type AppTemplate = "react-app";

/**
 * This error is thrown when input validation fails.
 */
export class InputValidationError extends Error {
    public readonly type = "IVE";

    public constructor(public readonly inputName: string, public readonly value: string, message: string) {
        super(message);

        // restore prototype chain
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

/**
 * Type guard testing whether object is a type of {@link InputValidationError}.
 */
export function isInputValidationError(obj: unknown): obj is InputValidationError {
    return !isEmpty(obj) && (obj as InputValidationError).type === "IVE";
}
