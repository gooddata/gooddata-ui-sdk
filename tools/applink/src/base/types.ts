// (C) 2020-2024 GoodData Corporation

/**
 * Descriptor for package sources
 */
export type SourceDescriptor = {
    /**
     * Absolute path to the root directory where SDK is located.
     */
    root: string;

    /**
     * Mapping of package name to SDK Package
     */
    packages: Record<string, PackageDescriptor>;

    /**
     * Mapping of package dir (relative to SDK root) to SDK Package.
     */
    packagesByDir: Record<string, PackageDescriptor>;

    /**
     * Inter-package dependency graph
     */
    dependencyGraph: DependencyGraph;
};

/**
 * SDK Package descriptor
 */
export type PackageDescriptor = {
    /**
     * Package name (e.g. `@gooddata/api-client-tiger`)
     */
    packageName: string;

    /**
     * Absolute path to the package directory.
     */
    directory: string;

    /**
     * Package's installation directory, split to segments (e.g. `['@gooddata', 'api-client-tiger']`)
     */
    installDir: string[];

    /**
     * Package's package.json - as read directly from the package's directory.
     */
    packageJson: PackageJson;

    /**
     * Package metadata from `rush.json`.
     */
    rushPackage: RushPackageDescriptor;
};

/**
 * Rush package metadata
 */
export type RushPackageDescriptor = {
    /**
     * Package name
     */
    packageName: string;

    /**
     * Relative path to the package folder (libs/api-client-tiger)
     */
    projectFolder: string;

    /**
     * 3rd party dependency review category
     */
    reviewCategory: string;

    /**
     * Versioning policy name
     */
    versionPolicyName: string;

    /**
     * Indicates whether package should be published.
     */
    shouldPublish: boolean;
};

/**
 * Minimalistic typing for the package json. Contains just the properties that applink needs.
 */
export type PackageJson = {
    version: string;
    scripts?: Record<string, string>;
    dependencies?: Record<string, string>;
    files?: string[];
};

/**
 * Dependency graph between packages
 */
export type DependencyGraph = {
    /**
     * All nodes in the graph.
     */
    nodes: string[];

    /**
     * Nodes in a set for quick lookups.
     */
    nodesSet: Set<string>;

    /**
     * All edges - dependencies between packages.
     */
    edges: Dependency[];

    /**
     * Outgoing edges for each node. E.g. edges grouped by 'from'.
     */
    outgoing: Record<string, Dependency[]>;

    /**
     * Incoming edges for each node. E.g. edges grouped by 'to'.
     */
    incoming: Record<string, Dependency[]>;
};

export type DependencyType = "prod" | "dev" | "peer";

export const AllDepdencyTypes: DependencyType[] = ["prod", "dev", "peer"];

/**
 * Dependency between SDK packages.
 */
export type Dependency = {
    /**
     * Package which has dependency.
     */
    from: string;

    /**
     * The dependent package
     */
    to: string;

    /**
     * Type of dependency.
     */
    type: DependencyType;
};

export type TargetDescriptor = {
    /**
     * Absolute path to the directory in which the target resides.
     */
    root: string;

    /**
     * All dependencies on source packages which were found in the target's node_modules
     */
    dependencies: TargetDependency[];
};

/**
 * Describes application's dependency on an SDK library.
 */
export type TargetDependency = {
    /**
     * Absolute path to the directory within app's node_modules where the SDK dependency is installed
     */
    directory: string;

    /**
     * The version of installed dependency.
     */
    version: string;

    /**
     * Package JSON of the installed package.
     */
    packageJson: PackageJson;

    /**
     * Full information about the source SDK package on which the app depends.
     */
    pkg: PackageDescriptor;
};
