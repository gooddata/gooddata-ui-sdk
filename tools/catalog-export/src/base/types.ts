// (C) 2007-2024 GoodData Corporation
import isEmpty from "lodash/isEmpty.js";

/**
 * This exception is thrown when a fatal error occurs during the export processing - be it during interfacing with
 * the backend or other unexpected errors. The exception contains 'rc' field - this specifies the exit code
 * for the catalog export process.
 *
 * The CatalogExportError is produced in 'log-and-throw' fashion. Code that raises the exception SHOULD do all the
 * necessary logging and cleanup and then throw the CatalogExportError. The top-level error handler then emits
 * message included in the error and exits process with exit code equal to the included `rc` field.
 */
export class CatalogExportError extends Error {
    constructor(message: string, public readonly rc: number) {
        super(message);
    }
}

/**
 * Exporter configuration
 */
export type CatalogExportConfig = {
    /**
     * Hostname where project lives
     */
    hostname: string | null;

    /**
     * Identifier of the workspace
     */
    workspaceId: string | null;

    /**
     * Tiger API Token for authentication
     */
    token: string | null;

    /**
     * File to write output to.
     */
    catalogOutput: string | null;
};

export function getConfiguredWorkspaceId(config: CatalogExportConfig): string | null {
    if (config.workspaceId) {
        return config.workspaceId;
    }

    return null;
}

/**
 * Subset of object meta used in exporter
 */
export type ObjectMeta = {
    /**
     * Object's title
     */
    title: string;

    /**
     * Object's identifier
     */
    identifier: string;

    /**
     * Object's tags
     */
    tags: string;
};

export type Attribute = {
    attribute: {
        meta: ObjectMeta;
        content: {
            displayForms: DisplayForm[];
        };
    };
};

export type DisplayForm = {
    meta: ObjectMeta;
};

export type Metric = {
    metric: {
        meta: ObjectMeta;
    };
};

export type Fact = {
    fact: {
        meta: ObjectMeta;
    };
};

export type DateDataSet = {
    dateDataSet: {
        meta: ObjectMeta;
        content: {
            attributes: Attribute[];
        };
    };
};

export type Catalog = {
    attributes: Attribute[];
    metrics: Metric[];
    facts: Fact[];
};

export type WorkspaceMetadata = {
    workspaceId: string;
    catalog: Catalog;
    dateDataSets: DateDataSet[];
    insights: ObjectMeta[];
    analyticalDashboards: ObjectMeta[];
};

//
// Type Guards
//

export function isAttribute(obj: unknown): obj is Attribute {
    return !isEmpty(obj) && (obj as Attribute).attribute !== undefined;
}

export function isMetric(obj: unknown): obj is Metric {
    return !isEmpty(obj) && (obj as Metric).metric !== undefined;
}

export function isFact(obj: unknown): obj is Fact {
    return !isEmpty(obj) && (obj as Fact).fact !== undefined;
}

export function isCatalogExportError(obj: unknown): obj is CatalogExportError {
    return !isEmpty(obj) && (obj as CatalogExportError).rc !== undefined;
}
