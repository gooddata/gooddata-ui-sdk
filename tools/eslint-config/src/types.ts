// (C) 2025 GoodData Corporation

export type Rules<Prefix extends string = ""> = {
    [key in `${Prefix}${string}`]: string | number | object | object[];
};

interface IPackage {
    name: string;
    version: string;
}

interface IConfigurationBase {
    /**
     * Packages which are needed for this configuration
     */
    packages?: IPackage[];
}

interface IConfigurationShared<RulePrefix extends string> {
    /**
     * Parser to be set when this configuration is active
     */
    parser?: string;
    /**
     * Plugin to add for this configuration to work
     */
    plugin?: string;
    /**
     * Configurations to extend as port of this configuration
     */
    extends?: string[];
    /**
     * Parser options to be set when this configuration is active
     */
    parserOptions?: Record<string, number | string>;
    /**
     * Rules for this configuration
     */
    rules?: Rules<RulePrefix>;
    /**
     * Settings to be added when this configuration is active
     */
    settings?: Record<string, object>;
    /**
     * Env values to be set when this configuration is active
     */
    env?: Record<string, boolean>;
    /**
     * Ignore patterns to be added when this configuration is active
     */
    ignorePatterns?: string[];
}

interface IOverride<RulePrefix extends string> extends IConfigurationShared<RulePrefix> {
    files: string[];
}

export interface IConfiguration<RulePrefix extends string = "">
    extends IConfigurationBase,
        IConfigurationShared<RulePrefix> {
    override?: IOverride<RulePrefix>;
}
