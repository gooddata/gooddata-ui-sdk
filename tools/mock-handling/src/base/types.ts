// (C) 2007-2024 GoodData Corporation
import isEmpty from "lodash/isEmpty.js";

/**
 * This exception is thrown when a fatal error occurs during the data recording.
 *
 * The DataRecorderError is produced in 'log-and-throw' fashion. Code that raises the exception SHOULD do all the
 * necessary logging and cleanup and then throw the DataRecorderError. The top-level error handler then emits
 * message included in the error and exits process with exit code equal to the included `rc` field.
 */
export class DataRecorderError extends Error {
    constructor(message: string, public readonly rc: number) {
        super(message);
    }
}

export type BackendType = "bear" | "tiger";

/*
 * Defines types used across catalog exporter
 */

/**
 * Exporter configuration
 */
export type DataRecorderConfig = {
    /**
     * Hostname where project lives
     */
    hostname: string | null;

    /**
     * Identifier of the project
     */
    projectId: string | null;

    /**
     * Tiger token
     */
    tigerToken: string | null;

    /**
     * Directory with recordings inputs & outputs.
     */
    recordingDir: string | null;

    /**
     * Backend type: bear or tiger.
     */
    backend: BackendType | null;

    /**
     * If specified, projectId will be replaced with this value in all files written by
     * the mock handling tool.
     */
    replaceProjectId: string | null;
};

export function isDataRecorderError(obj: unknown): obj is DataRecorderError {
    return !isEmpty(obj) && (obj as DataRecorderError).rc !== undefined;
}
