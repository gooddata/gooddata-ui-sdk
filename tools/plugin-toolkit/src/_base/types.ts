// (C) 2021-2024 GoodData Corporation

import { OptionValues } from "commander";

/**
 * Program & command level options are grabbed from the commander and sent together to the action.
 */
export type ActionOptions = {
    /**
     * Program-level options. These are the top level options defined directly using `program.option(...)`
     */
    programOpts: OptionValues;

    /**
     * Command-level options. These are options defined via `program.command(..).option(..)`
     */
    commandOpts: OptionValues;
};

/**
 * Supported languages to use in plugin projects bootstrapped by the app.
 */
export type TargetAppLanguage = "ts" | "js";

/**
 * Supported package managers that the tool can use to auto-install dependencies for the boostrapped projects.
 */
export type SupportedPackageManager = "npm" | "yarn";

/**
 * This error is thrown when input validation fails.
 */
export class InputValidationError extends Error {
    public readonly type = "IVE";

    public constructor(public readonly inputName: string, public readonly value: string, message: string) {
        super(message);

        // restore prototype chain
        Object.setPrototypeOf(this, new.target.prototype);
    }
}

/**
 * Type guard testing whether object is a type of {@link InputValidationError}.
 */
export function isInputValidationError(obj: Error): obj is InputValidationError {
    return obj && (obj as InputValidationError).type === "IVE";
}
