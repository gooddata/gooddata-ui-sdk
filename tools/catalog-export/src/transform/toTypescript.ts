// (C) 2007-2024 GoodData Corporation
import {
    ImportDeclarationStructure,
    OptionalKind,
    Project,
    SourceFile,
    VariableDeclarationKind,
    VariableStatementStructure,
} from "ts-morph";
import { MeasureAggregation } from "@gooddata/sdk-model";
import { Attribute, DateDataSet, DisplayForm, Fact, Metric, WorkspaceMetadata } from "../base/types.js";
import { createUniqueVariableName, stringToVariableName, TakenNamesSet } from "./titles.js";

export type TypescriptOutput = {
    project: Project;
    sourceFile: SourceFile;
};

//
// Constants
//

const FILE_DIRECTIVES = ["/* eslint-disable */"];
const FILE_HEADER = `/* THIS FILE WAS AUTO-GENERATED USING CATALOG EXPORTER; YOU SHOULD NOT EDIT THIS FILE; GENERATE TIME: ${new Date().toISOString()}; */`;
const INSIGHT_MAP_VARNAME = "Insights";
const ANALYTICAL_DASHBOARD_MAP_VARNAME = "Dashboards";
const DATE_DATASETS_MAP_VARNAME = "DateDatasets";
// "count" aggregation is not allowed for Facts, so do not generate code for that
const FACT_AGGREGATIONS: MeasureAggregation[] = ["sum", "avg", "min", "max", "median", "runsum"];

//
// Variable naming support & strategies
//

let GlobalNameScope: TakenNamesSet = {};

type NamingStrategy = (title: string, scope: TakenNamesSet) => string;

type NamingStrategies = {
    dataSetProperty: NamingStrategy;
    dataSetAttributeProperty: NamingStrategy;
    attribute: NamingStrategy;
    displayForm: NamingStrategy;
};

const DefaultNaming: NamingStrategies = {
    dataSetProperty: uniqueVariable,
    dataSetAttributeProperty: uniqueVariable,
    attribute: uniqueVariable,
    displayForm: uniqueVariable,
};

/**
 * This is a wrapper on top of createUniqueVariableName() which mutates the input name scope - it enters
 * the variable name into the scope.
 *
 * @param title - md object title
 * @param nameScope - scope containing already taken variable names, defaults to global scope
 */
function uniqueVariable(title: string, nameScope: TakenNamesSet = GlobalNameScope): string {
    const variableName = createUniqueVariableName(title, nameScope);
    nameScope[variableName] = true;

    return variableName;
}

//
// Transformation to ts-morph structures
//

function initialize(outputFile: string): TypescriptOutput {
    const project = new Project({});
    const sourceFile = project.createSourceFile(
        outputFile,
        {
            leadingTrivia: [...FILE_DIRECTIVES, FILE_HEADER],
        },
        { overwrite: true },
    );

    return {
        project,
        sourceFile,
    };
}

function generateSdkModelImports(): OptionalKind<ImportDeclarationStructure> {
    return {
        moduleSpecifier: "@gooddata/sdk-model",
        namedImports: ["newAttribute", "newMeasure", "IAttribute", "IMeasure", "IMeasureDefinition", "idRef"],
        leadingTrivia:
            "// @ts-ignore ignore unused imports here if they happen (e.g. when there is no measure in the workspace)",
    };
}

function generateAttributeDisplayForm(
    displayForm: DisplayForm,
    attributeVariableName: string,
    nameScope: TakenNamesSet,
    naming: NamingStrategies,
): string {
    const { meta } = displayForm;
    const dfVariableName = naming.displayForm(meta.title, nameScope);
    let variableName = attributeVariableName === dfVariableName ? "Default" : dfVariableName;

    if (variableName.startsWith(attributeVariableName)) {
        // we need to do the variable conversion again as it can happen that after removing the attribute name what is left starts with a number
        // for example Foo attribute and Foo123Bar display form becomes 123Bar which is invalid in this context
        variableName = stringToVariableName(variableName.substr(attributeVariableName.length));
    }

    return `/** \n* Display Form Title: ${meta.title}  \n* Display Form ID: ${meta.identifier}\n*/\n${variableName}: newAttribute('${meta.identifier}')`;
}

/**
 * Generates display form property initializers for an attribute. When naming the display form properties, the naming strategy will
 * be used; the name returned from the strategy will be further tweaked to improve readability of the generated mapping:
 *
 * - if it equals to name of attribute, the display form key will be 'Default' (indicating default display form, preventing
 *   weird mapping of say Product.Product)
 *
 * - if the label is prefixed with attribute name, the prefix will be trimmed (going from say Product.ProductName to Product.Name)
 *
 * @param attribute - attribute to work with
 * @param attributeVariableName - variable name assigned to the attribute
 * @param naming - naming strategy to use
 */
function generateDisplayFormPropertyInitializers(
    attribute: Attribute,
    attributeVariableName: string,
    naming: NamingStrategies = DefaultNaming,
): string[] {
    const localNameScope: TakenNamesSet = {};
    return attribute.attribute.content.displayForms.map((df) =>
        generateAttributeDisplayForm(df, attributeVariableName, localNameScope, naming),
    );
}

/**
 * Generates attribute constant. Works as follows:
 *
 * - If the attribute has single display form, generates a constant of DfTitle ⇒ newAttribute(id)
 * - If the attribute has multiple display forms, generates a constant that is an object mapping different
 *   DfTitles ⇒ newAttribute(dfId)
 *
 * **NOTE**: this function has side effect. as it generates the statement (which it returns) it will update the
 * attribute instance with name of the generated constant.
 *
 * @param attribute - attribute to generate definitions for
 * @param naming - naming scope to ensure variable name uniqueness
 * @param deprecation - if specified, add `@deprecated` doc with this message
 */
function generateAttributeConstant(
    attribute: Attribute,
    naming: NamingStrategies = DefaultNaming,
): OptionalKind<VariableStatementStructure> {
    const { meta } = attribute.attribute;
    const variableName = naming.attribute(meta.title, GlobalNameScope);
    const { displayForms } = attribute.attribute.content;
    const comments = [`Attribute Title: ${meta.title}`, `Attribute ID: ${meta.identifier}`];

    const docs = [comments.join("\n")];

    if (displayForms.length === 1) {
        /*
         * If there is a single DF for the attribute, then have const AttrName = newAttribute(...)
         */
        const displayForm = displayForms[0];

        return {
            declarationKind: VariableDeclarationKind.Const,
            isExported: true,
            docs,
            declarations: [
                {
                    name: variableName,
                    type: "IAttribute",
                    initializer: `newAttribute('${displayForm.meta.identifier}')`,
                },
            ],
        };
    } else {
        /*
         * If there are multiple DFs, have mapping of const AttrName = { DfName: newAttribute(), OtherDfName: newAttribute()}
         */
        const displayFormInits: string[] = generateDisplayFormPropertyInitializers(
            attribute,
            variableName,
            naming,
        );

        return {
            declarationKind: VariableDeclarationKind.Const,
            isExported: true,
            docs,
            declarations: [
                {
                    name: variableName,
                    initializer: `{ ${displayFormInits.join(",")} }`,
                },
            ],
        };
    }
}

function generateAttributes(
    projectMeta: WorkspaceMetadata,
): ReadonlyArray<OptionalKind<VariableStatementStructure>> {
    return projectMeta.catalog.attributes.map((a) => generateAttributeConstant(a));
}

function generateMeasureFromMetric(metric: Metric): OptionalKind<VariableStatementStructure> {
    const { meta } = metric.metric;
    const variableName = uniqueVariable(meta.title);

    return {
        declarationKind: VariableDeclarationKind.Const,
        isExported: true,
        docs: [`Metric Title: ${meta.title}\nMetric ID: ${meta.identifier}\nMetric Type: MAQL Metric`],
        declarations: [
            {
                name: variableName,
                type: "IMeasure<IMeasureDefinition>",
                initializer: `newMeasure(idRef('${meta.identifier}', "measure"))`,
            },
        ],
    };
}

function generateFactAggregations(fact: Fact): string[] {
    const { meta } = fact.fact;

    return FACT_AGGREGATIONS.map((aggregation) => {
        const jsDoc = `/** \n* Fact Title: ${meta.title}  \n* Fact ID: ${meta.identifier}\n * Fact Aggregation: ${aggregation}\n*/`;
        const name = aggregation.charAt(0).toUpperCase() + aggregation.substr(1);

        return `${jsDoc}\n${name}: newMeasure(idRef('${meta.identifier}', "fact"), m => m.aggregation('${aggregation}'))`;
    });
}

function generateMeasuresFromFacts(fact: Fact): OptionalKind<VariableStatementStructure> {
    const { meta } = fact.fact;
    const variableName = uniqueVariable(meta.title);

    const aggregationInits: string[] = generateFactAggregations(fact);

    return {
        declarationKind: VariableDeclarationKind.Const,
        isExported: true,
        docs: [`Fact Title: ${meta.title}\nFact ID: ${meta.identifier}`],
        declarations: [
            {
                name: variableName,
                initializer: `{ ${aggregationInits.join(",")} } `,
            },
        ],
    };
}

/**
 * Generates simple measures from metrics and facts defined in the project.
 *
 * - For metrics (MAQL stuff) this will generate constant with MeasureTitle = newMeasure(id)
 * - For facts, this will generate constant initialized to an object which defines all possible
 *   aggregations of the measure.
 */
function generateMeasures(
    projectMeta: WorkspaceMetadata,
): ReadonlyArray<OptionalKind<VariableStatementStructure>> {
    const fromMetrics = projectMeta.catalog.metrics.map(generateMeasureFromMetric);
    const fromFacts = projectMeta.catalog.facts.map(generateMeasuresFromFacts);

    return fromMetrics.concat(fromFacts);
}

function generateDateDataSetAttributes(dd: DateDataSet, naming: NamingStrategies): string[] {
    const ddScope = {};

    return dd.dateDataSet.content.attributes.map((a) => {
        const { meta } = a.attribute;
        const { title, identifier } = meta;

        const attributePropertyName = naming.dataSetAttributeProperty(meta.title, ddScope);
        const jsDoc = `/** \n* Date Attribute: ${title}  \n* Date Attribute ID: ${identifier}\n*/`;
        const displayFormProps = generateDisplayFormPropertyInitializers(a, attributePropertyName, naming);

        return `${jsDoc}\n${attributePropertyName}: { ref: idRef('${identifier}', 'attribute'), identifier: '${identifier}', ${displayFormProps.join(
            ",",
        )} }`;
    });
}

function generateDateDataSetInitialzier(
    dd: DateDataSet,
    naming: NamingStrategies,
    namingScope: TakenNamesSet,
): string {
    const { title, identifier } = dd.dateDataSet.meta;
    const propName = naming.dataSetProperty(title, namingScope);
    const jsDoc = `/** \n* Date Data Set Title: ${title}  \n* Date Data Set ID: ${identifier}\n*/`;
    const attributes = generateDateDataSetAttributes(dd, naming);

    return `${jsDoc}\n${propName}: { ref: idRef('${identifier}', 'dataSet'), identifier: '${identifier}', ${attributes.join(
        ", ",
    )} }`;
}

/**
 * Given date datasets, this generates a DateDataSets constants which is an object with a property per-dataset. For each
 * data set, there is another object with: `ref` property which contains data set's idRef and then property per-attribute.
 *
 * These per-attribute properties reference existing constants generated for the date attributes ⇒ this function MUST
 * be called after generateAttributes.
 *
 * @param dds - date data sets
 * @param naming - naming strategy
 */
function generateDateDataSetMapping(
    dds: DateDataSet[],
    naming: NamingStrategies,
): OptionalKind<VariableStatementStructure> {
    const mappingScope = {};
    const ddMappings = dds.map((dd) => generateDateDataSetInitialzier(dd, naming, mappingScope));

    return {
        declarationKind: VariableDeclarationKind.Const,
        isExported: true,
        docs: ["Available Date Data Sets"],
        declarations: [
            {
                name: DATE_DATASETS_MAP_VARNAME,
                initializer: `{ ${ddMappings.join(", ")} }`,
            },
        ],
    };
}

function generateDateDataSets(
    projectMeta: WorkspaceMetadata,
): ReadonlyArray<OptionalKind<VariableStatementStructure>> {
    return [generateDateDataSetMapping(projectMeta.dateDataSets, DefaultNaming)];
}

/**
 * Declares a constant initialized to object mapping InsightTitle ⇒ insight identifier.
 *
 * @param projectMeta - project metadata containing the insights
 */
function generateInsights(projectMeta: WorkspaceMetadata): OptionalKind<VariableStatementStructure> {
    const insightInitializer: string[] = projectMeta.insights.map((insight) => {
        const propName = uniqueVariable(insight.title);
        const jsDoc = `/** \n* Insight Title: ${insight.title}  \n* Insight ID: ${insight.identifier}\n*/`;

        return `${jsDoc}\n${propName}: '${insight.identifier}'`;
    });

    return {
        declarationKind: VariableDeclarationKind.Const,
        isExported: true,
        declarations: [
            {
                name: INSIGHT_MAP_VARNAME,
                initializer: `{ ${insightInitializer.join(",")} }`,
            },
        ],
    };
}

/**
 * Declares a constant initialized to object mapping analyticalDashboard ⇒ analyticalDashboard identifier.
 *
 * @param projectMeta - project metadata containing the analyticalDashboards
 */
function generateAnalyticalDashboards(
    projectMeta: WorkspaceMetadata,
): OptionalKind<VariableStatementStructure> {
    const analyticalDashboardInitializer: string[] = projectMeta.analyticalDashboards.map((dashboard) => {
        const propName = uniqueVariable(dashboard.title);
        const jsDoc = `/** \n* Dashboard Title: ${dashboard.title}  \n* Dashboard ID: ${dashboard.identifier}\n*/`;

        return `${jsDoc}\n${propName}: '${dashboard.identifier}'`;
    });

    return {
        declarationKind: VariableDeclarationKind.Const,
        isExported: true,
        declarations: [
            {
                name: ANALYTICAL_DASHBOARD_MAP_VARNAME,
                initializer: `{ ${analyticalDashboardInitializer.join(",")} }`,
            },
        ],
    };
}

/**
 * Transforms project metadata into model definitions in TypeScript. The resulting TS source file will contain
 * constant declarations for the various objects in the metadata:
 *
 * - Attributes with single display form are transformed to constants initialized with respective newAttribute()
 * - Attributes with multiple display forms are transformed to constants initialized to map of df name ⇒ newAttribute()
 * - Metrics (MAQL) will be transformed to constants initialized with respective newMeasure()
 * - Metrics from facts will be transformed to constants initialized to map of aggregation ⇒ newMeasure()
 * - Date data set attributes will be transformed using the same logic as normal attributes, albeit with slightly
 *   modified variable naming strategy
 *
 * @param projectMeta - project metadata to transform to typescript
 * @param outputFile - output typescript file
 * @returns return of the transformation process, new file is not saved at this point
 */
export function transformToTypescript(projectMeta: WorkspaceMetadata, outputFile: string): TypescriptOutput {
    GlobalNameScope = {};

    const output = initialize(outputFile);
    const { sourceFile } = output;

    sourceFile.addImportDeclaration(generateSdkModelImports());
    sourceFile.addVariableStatements(generateAttributes(projectMeta));
    sourceFile.addVariableStatements(generateMeasures(projectMeta));
    sourceFile.addVariableStatements(generateDateDataSets(projectMeta));
    sourceFile.addVariableStatement(generateInsights(projectMeta));
    sourceFile.addVariableStatement(generateAnalyticalDashboards(projectMeta));

    return output;
}
