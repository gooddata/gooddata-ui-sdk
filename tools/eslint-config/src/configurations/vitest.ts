// (C) 2025 GoodData Corporation

import type { IConfiguration } from "../types.js";

const configuration: IConfiguration<"@vitest"> = {
    packages: [
        {
            name: "@vitest/eslint-plugin",
            version: "1.3.5",
        },
    ],
    plugin: "@vitest",
    extends: ["plugin:@vitest/legacy-recommended"],
    rules: {
        "@vitest/expect-expect": "off",
        "@vitest/no-commented-out-tests": "warn",
        "@vitest/valid-title": "warn",
        "@vitest/no-disabled-tests": "warn",
        "@vitest/no-focused-tests": "warn",
        "@vitest/no-identical-title": "warn",
        "@vitest/valid-expect": "warn",
    },
};

export default configuration;
