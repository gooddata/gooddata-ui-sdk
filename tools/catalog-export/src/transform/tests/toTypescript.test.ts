// (C) 2007-2024 GoodData Corporation
import { describe, it, expect } from "vitest";
import { transformToTypescript } from "../toTypescript.js";
import { WorkspaceMetadata } from "../../base/types.js";

function createObjMetaMock(title: string) {
    return {
        title,
        identifier: `identifier-${title}`,
        tags: `tags-${title}`,
    };
}

function createFactMock(title: string) {
    return {
        fact: {
            meta: createObjMetaMock(title),
        },
    };
}

function createMetricMock(title: string) {
    return {
        metric: {
            meta: createObjMetaMock(title),
        },
    };
}

function createAttributeMock(title: string) {
    return {
        attribute: {
            content: {
                displayForms: [],
            },
            meta: createObjMetaMock(title),
        },
    };
}

const projectMeta: WorkspaceMetadata = {
    workspaceId: "workspaceId",
    catalog: {
        attributes: [createAttributeMock("attribute1")],
        metrics: [createMetricMock("metric1"), createMetricMock("metric2")],
        facts: [createFactMock("fact1")],
    },
    dateDataSets: [],
    insights: [createObjMetaMock("insight1")],
    analyticalDashboards: [],
};

describe("transformToTypescript", () => {
    it("creates new catalog", async () => {
        const transformResult = transformToTypescript(await projectMeta, "testOutput.ts");

        /*
         * NOTE: source.getFullText() should not be used here as it includes the leading comments that
         * contains timestamp describing when was the file generated.
         */
        expect(transformResult.sourceFile.getText()).toMatchSnapshot();
    });
});
