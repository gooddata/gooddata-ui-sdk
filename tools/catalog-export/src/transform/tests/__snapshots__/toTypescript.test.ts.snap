// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformToTypescript > creates new catalog 1`] = `
"import { newAttribute, newMeasure, IAttribute, IMeasure, IMeasureDefinition, idRef } from "@gooddata/sdk-model";

/**
 * Attribute Title: attribute1
 * Attribute ID: identifier-attribute1
 */
export const Attribute1 = {  };
/**
 * Metric Title: metric1
 * Metric ID: identifier-metric1
 * Metric Type: MAQL Metric
 */
export const Metric1: IMeasure<IMeasureDefinition> = newMeasure(idRef('identifier-metric1', "measure"));
/**
 * Metric Title: metric2
 * Metric ID: identifier-metric2
 * Metric Type: MAQL Metric
 */
export const Metric2: IMeasure<IMeasureDefinition> = newMeasure(idRef('identifier-metric2', "measure"));
/**
 * Fact Title: fact1
 * Fact ID: identifier-fact1
 */
export const Fact1 = { /** 
    * Fact Title: fact1  
    * Fact ID: identifier-fact1
     * Fact Aggregation: sum
    */
    Sum: newMeasure(idRef('identifier-fact1', "fact"), m => m.aggregation('sum')),/** 
    * Fact Title: fact1  
    * Fact ID: identifier-fact1
     * Fact Aggregation: avg
    */
    Avg: newMeasure(idRef('identifier-fact1', "fact"), m => m.aggregation('avg')),/** 
    * Fact Title: fact1  
    * Fact ID: identifier-fact1
     * Fact Aggregation: min
    */
    Min: newMeasure(idRef('identifier-fact1', "fact"), m => m.aggregation('min')),/** 
    * Fact Title: fact1  
    * Fact ID: identifier-fact1
     * Fact Aggregation: max
    */
    Max: newMeasure(idRef('identifier-fact1', "fact"), m => m.aggregation('max')),/** 
    * Fact Title: fact1  
    * Fact ID: identifier-fact1
     * Fact Aggregation: median
    */
    Median: newMeasure(idRef('identifier-fact1', "fact"), m => m.aggregation('median')),/** 
    * Fact Title: fact1  
    * Fact ID: identifier-fact1
     * Fact Aggregation: runsum
    */
    Runsum: newMeasure(idRef('identifier-fact1', "fact"), m => m.aggregation('runsum')) } ;
/** Available Date Data Sets */
export const DateDatasets = {  };
export const Insights = { /** 
    * Insight Title: insight1  
    * Insight ID: identifier-insight1
    */
    Insight1: 'identifier-insight1' };
export const Dashboards = {  };
"
`;
